EMDS: Added documentation for EMDS (#8)
Balaklaka/sdk-nrf: doc/nrf/libraries/emds/emds.rst

.. _emds_readme:

Emergency data storage
######################

.. contents::
    :local:
    :depth: 2

> The emergency data storage (EMDS) library provides functionality that will prevent wear and tear of flash as data that are changing often are stored in memory, but can be stored in flash and be restored after reboot. It will store data within a guaranteed time, based on the amount of data that are being stored.
> EMDS can be used in situations of critical power shortage to store data in memory, e.g. before the battery that powers a device is depleted. It is important that the HW has appropriate functionality to sustain power long enough for storage to be completed before the device is fully discharged.

Reviewer: ```suggestion
The emergency data storage (EMDS) library provides persistent storage functionality that is designed to prevent wear and tear of flash memory. It is intended for storing data that undergoes frequent changes during runtime. To prevent frequent writes to the flash memory, the EMDS library will only write data to flash at the point when the device is shuting down. At reboot the module will restore this data. EMDS can store data within a guaranteed time, based on the amount of data that are being stored.
EMDS can be used in situations of critical power shortage to store data in memory, e.g. before the battery that powers a device is depleted. It is important that the HW has appropriate functionality to sustain power long enough for storage to be completed before the device is fully discharged.
```

> .. _emds_readme:

Reviewer 2: I'd suggest to change `memory` on `entry` (with clarification of what entry is) in all places when entry is assumed. Currently, it is difficult to differentiate which memory the document is talking about.

Emergency data storage
######################

.. contents::
    :local:
    :depth: 2

> The emergency data storage (EMDS) library provides functionality for storing of data with a guaranteed time, based on the amount of data that are being stored.
> It can be used in situations of critical power shortage to store data in memory before the battery that powers the device is about to fully discharge.

Reviewer: ```suggestion
The emergency data storage (EMDS) library provides functionality for storing of data within a guaranteed time, based on the amount of data that are being stored.
It can be used in situations of critical power shortage to store data in memory, e.g. before the battery that powers a device is depleted.
```

> 

Reviewer: I think it is wise to add a warning that informs the user that using this feature without the appropriate HW is likely to cause loss of data if a device abruptly loses power.

Author: This has been added in top, by saying HW should have func for sustain power.

.. note::
    The Power-fail comparator for the nRF528xx can not be used with this feature, as it will prevents the NVMC from performing write operations to the flash.

Overview
********

> The :kconfig:`CONFIG_EMDS` enables the emergency data storage. The application must initialize by using :ref:`emds_init`, the pre-allocated flash area used to store data in.
> The :kconfig:`CONFIG_EMDS_SECTOR_COUNT` defines how many sectors that should be used to store data. The space set aside shall be large than all the data stored. The larger value will increase the flash spaced used, but will reduce the wear of the flash area.
> This has to be balanced between amount of data stored, flash available and how of it will call the :ref:`emds_store`. It should not be necessary to have a large value, as there should be a limited set of data stored this way, as time is of essence here.

Reviewer: ```suggestion
The :kconfig:`CONFIG_EMDS` enables the emergency data storage. The application must initialize the pre-allocated flash area by using :ref:`emds_init`.
The :kconfig:`CONFIG_EMDS_SECTOR_COUNT` defines how many sectors that should be used to store data. The allocated storage space must be larger than the combined data stored by the application. Choosing a larger flash area will demand more resources, while simutaniously increase the life expectancy of the flash. The choosen size shuold reflect the amount of data stored, the available flash resources, and how the application calls the :ref:`emds_store`. In general it should not be necessary to allocate a large flash area, since there should be a limited set of data stored to ensure that writing the flash on shutdown completes swiftly.
```

The memory location that is going to be stored need to be added on initialization. All memory areas need to be provided through the :ref:`emds_entry_add` and :ref:`EMDS_STATIC_ENTRY_DEFINE`.

> When adding a memory area that should be stored when the emergency data storage is triggered it is important the each entry has it's own unique id that can be recreated after a reboot.

Reviewer 2: ```suggestion
When adding a memory area that should be stored when the emergency data storage is triggered it is important the each entry has its own unique id that is not changed after a reboot.
```

> The memory location that is going to be stored need to be added on initialization. All memory areas need to be provided through the :ref:`emds_entry_add` and :ref:`EMDS_STATIC_ENTRY_DEFINE`.
> When adding a memory area that should be stored when the emergency data storage is triggered it is important the each entry has it's own unique id that can be recreated after a reboot.

Reviewer: ```suggestion
The memory location that is going to be stored must be added on initialization. All memory areas must be provided through the :ref:`emds_entry_add` and :ref:`EMDS_STATIC_ENTRY_DEFINE`.
When adding a memory area that should be stored when the emergency data storage is triggered, it is important that each entry has it's own unique id that can be restored after a reboot.
```

The :ref:`emds_entry_load` will restore the data into the memory areas from the flash. This has to be done after all the memory areas has been added.

> After the data has been restored the application shall run a :ref:`emds_prepare`. This to get the pre allocated flash area prepared for receiving data, when the emergency data storage is needed. Available flash area must be larger than all data stored in the emergency data storage. If multiple store has been run and the flash area that is left is smaller than the needed data size, the flash area will be automatically erased to increase the empty flash area available.

Reviewer: ```suggestion
After restoring previous data the application must run a :ref:`emds_prepare` to prepare the flash area for receiving new data. If the remaining empty flash area is smaller than the required data size, the flash area will be automatically erased to increase the available flash area.
```

> The storage will be done in a deterministic time so it will be possible know how long it takes to stored all the registered data. However this is dependent on the chip so to measure time is important and :kconfig:`EMDS_FLASH_TIME_BASE_OVERHEAD_US`, :kconfig:`EMDS_FLASH_TIME_ENTRY_OVERHEAD_US` and :kconfig:`EMDS_FLASH_TIME_BASE_OVERHEAD_US` can be configured. It is important when :kconfig:`EMDS_FLASH_TIME_BASE_OVERHEAD_US` is added time the worst time is considered, that including time for erase cycles when doing garbage collection.

Reviewer 2: Which garbage collection are you talking about?

> The storage will be done in a deterministic time so it will be possible know how long it takes to stored all the registered data. However this is dependent on the chip so to measure time is important and :kconfig:`EMDS_FLASH_TIME_BASE_OVERHEAD_US`, :kconfig:`EMDS_FLASH_TIME_ENTRY_OVERHEAD_US` and :kconfig:`EMDS_FLASH_TIME_BASE_OVERHEAD_US` can be configured. It is important when :kconfig:`EMDS_FLASH_TIME_BASE_OVERHEAD_US` is added time the worst time is considered, that including time for erase cycles when doing garbage collection.

Reviewer: I find parts of this section somewhat confusing. I guess the GBC you mention is reffering to if you simoutaniously use NVS? In that case I think you need to elaborate a bit, since it could be interpreted as the EMDS use GBC.

Author: I believe this is now rewritten.

> This will also prevent wear and tear of flash as data that are changing often are stored in memory, but can be stored in flash and be restored after reboot.

Reviewer: I don't see the link between this paragraph and the section before. Does the configuration of `EMDS_FLASH_TIME_BASE_OVERHEAD_US` impact wear and tear of the flash? Or is it simply a statment that EMDS will reduce wear and tear compared to e.g. NVS? If the second situation is the case I suggest that this part is moved to the initial part of the Doc.

Author: This has been  moved.

> The application need to call :ref:`emds_store` to store all the RAM areas provided. This can be only be done once before it need to run :ref:`emds_prepare` again. When the data storage is complete a callback will be called if provided in :ref:`emds_init`. This to notify to the application that the data storage is complete. This can be used to reboot the cpu or any other function that maybe needed.
> It will also make the :ref:`emds_is_ready` function call return false, as it is no longer can guaranty that the storage will fit in the amount of flash space left.

Reviewer: ```suggestion
Before shutdown the application must call :ref:`emds_store` to store all the RAM areas provided. This can be only be done once before it need to run :ref:`emds_prepare` again. When the data storage is complete a callback will be called if provided in :ref:`emds_init`. This callback notifies the application that the data storage is complete, and can be used to reboot the cpu or execute other function that is needed.
After completion of :ref:`emds_store` the :ref:`emds_is_ready` function call will return false, since it can no longer guarantee that the data will fit in the flash area.
```

This can be used by the application detects that the power is about to be lost, call the :ref:`emds_store` and the emergency data store process will be initiated in a separate thread. It will then store all the registered memory areas.

> The :ref:`emds_is_ready` can be called to check if EMDS is prepared to store date. The :ref:`emds_store` can only be run once, and area has to be prepared before it can be used again.

Reviewer: Is it neccesary to add this here? Is it not sufficient that this is covered in the API refference?

Author: It is not directly written in the API ref.

> This can be used by the application detects that the power is about to be lost, call the :ref:`emds_store` and the emergency data store process will be initiated in a separate thread. It will then store all the registered memory areas.
> The :ref:`emds_is_ready` can be called to check if EMDS is prepared to store date. The :ref:`emds_store` can only be run once, and area has to be prepared before it can be used again.

Reviewer: ```suggestion
When invoked, :ref:`emds_store` will trigger the emergency data store process in a separate thread. It will then store all the registered memory areas.
Invocation of this call should be performed when the application detects a loss of power, when a reboot is triggered, etc. 
The :ref:`emds_is_ready` can be called to check if EMDS is prepared to store data. The :ref:`emds_store` can only be run once, and area has to be prepared before it can be used again.
```

> 

Reviewer: Some sort of diagram would really make it easier to understand the flow of function calls during initialization, shutdown and reboot.

Author: Sequence diagram has been added, and configuration description been added.

.. note::
    Before the application call the :ref:`emds_store` the application should try to shutdown application specific features that are using a lot of power.
    Shutting down these features may save power and prolong the time the cpu is alive, and speed up the the storage time.

>     For example, if Bluetooth is used, disabling Bluetooth will save power and stopping mpsl scheduler will increase storage speed.

Reviewer 2: IMHO It is worth to add example of how to do that.

Author: This is done in the test (emds_api). And I was thinking of adding that to the mesh sample when doing that.

Reviewer 2: then the reference on the sample should be here

Author: There is no sample to reference, only a test.

Reviewer 2: Does it have the sense to provide reference on the test as 'sample'?

Author: Removed the reference to the test.

>     Before the application call the :ref:`emds_store` the application should try to shutdown application specific features that are using a lot of power.
>     Shutting down these features may save power and prolong the time the cpu is alive, and speed up the the storage time.
>     For example, if Bluetooth is used, disabling Bluetooth will save power and stopping mpsl scheduler will increase storage speed.

Reviewer: ```suggestion
    Before the application call the :ref:`emds_store` the application should try to shutdown application specific features that consumes a lot of power.
    Shutting down these features may prolong the time the cpu is alive, and improve the storage time.
    For example, if Bluetooth is used, disabling Bluetooth before shutdown will save power and stopping mpsl scheduler will increase storage speed.
```
